File: surf-api-core/surf-api-core-api/src/main/java/dev/slne/surf/surfapi/core/api/util/LocationFactory.java
package dev.slne.surf.surfapi.core.api.util;

import com.github.retrooper.packetevents.protocol.world.Location;
import org.jetbrains.annotations.ApiStatus;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;
import org.spongepowered.math.GenericMath;

public interface LocationFactory {

  static double distanceSquared(@NotNull Location locationA, @NotNull Location locationB) {
    return square(locationA.getX() - locationB.getX()) + square(locationA.getY() - locationB.getY())
        + square(locationA.getZ() - locationB.getZ());
  }

  static double distance(@NotNull Location locationA, @NotNull Location locationB) {
    return GenericMath.sqrt(distanceSquared(locationA, locationB));
  }

  @ApiStatus.Internal
  @Contract(pure = true)
  static double square(double value) {
    return value * value;
  }
}

File: surf-api-core/surf-api-core-api/src/main/kotlin/dev/slne/surf/surfapi/core/api/util/LocationFactory.kt
package dev.slne.surf.surfapi.core.api.util

import com.github.retrooper.packetevents.protocol.world.Location
import org.spongepowered.math.GenericMath

@Deprecated("Not longer maintained.")
interface LocationFactory {
    companion object {
        @JvmStatic
        @Deprecated("Not longer maintained.", ReplaceWith("locationA.position.distanceSquared(locationB.position)"))
        fun distanceSquared(locationA: Location, locationB: Location): Double {
            return locationA.position.distanceSquared(locationB.position)
        }

        @JvmStatic
        @Deprecated("Not longer maintained.", ReplaceWith("locationA.position.distance(locationB.position)"))
        fun distance(locationA: Location, locationB: Location): Double {
            return GenericMath.sqrt(locationA.position.distanceSquared(locationB.position))
        }
    }
}

@Deprecated("Not longer maintained.", ReplaceWith("this.position.distanceSquared(location.position)"))
infix fun Location.distanceSquared(location: Location): Double {
    return this.position.distanceSquared(location.position)
}

@Deprecated("Not longer maintained.", ReplaceWith("this.position.distance(location.position)"))
infix fun Location.distance(location: Location): Double {
    return this.position.distance(location.position)
}
